src/app/gui/queryresults/queryresultsservice.js

```diff
@@ -1,1874 +1,2555 @@
-    accumulator[property] = {};
-    return accumulator;
-  }, {});
-  layer.features.map((feature, index)=> {
-    Object.keys(propertiesObject).forEach(property =>{
-      propertiesObject[property][index] = null;
+  /**
+   * @FIXME add description
+   *
+   * @param layer
+   */
+  showLayerDownloadFormats(layer) {
+    const name = DownloadFormats.name;
+    layer[name].active = !layer[name].active;
+    this.setLayerActionTool({
+      layer,
+      component: layer[name].active ? DownloadFormats : null,
+      config: layer[name].active ? this.state.actiontools[name][layer.id] : null
     })
-  });
-  return Vue.observable(propertiesObject);
-};
+  }
 
-/**
- * Method to get action referred to layer getting the acion id
- * @param layer layer linked to action
- * @param id action id
- * @returns {*}
- */
-proto.getActionLayerById = function({layer, id}={}){
-  return this.state.layersactions[layer.id].find(action => action.id === id);
-};
+  /**
+   * @FIXME add description
+   *
+   * @param type
+   * @param layer
+   * @param features
+   * @param action
+   * @param index
+   */
+  async downloadFeatures(type, layer, features = [], action, index) {
 
-/**
- * Set current layer action tool in feature
- * @param layer current layer
- * @param index feature index
- * @param value component value or null
- */
-proto.setCurrentActionLayerFeatureTool = function({layer, action, index, component=null}={}){
-  if (component){
-    if (this.state.currentactiontools[layer.id][index] && action.id !== this.state.currentactionfeaturelayer[layer.id][index].id && this.state.currentactionfeaturelayer[layer.id][index].toggleable)
-      this.state.currentactionfeaturelayer[layer.id][index].state.toggled[index] = false;
-    this.state.currentactionfeaturelayer[layer.id][index] = action;
-  } else this.state.currentactionfeaturelayer[layer.id][index] = null;
-  this.state.currentactiontools[layer.id][index] = component;
-};
+    if (features && !Array.isArray(features)) {
+      features = [features];
+    }
+
+    const { query = {} } = this.state;
+    const data           = {
+      fids: features.map(f => f.attributes[G3W_FID]).join(',')
+    };
+
+    /**
+     * A function that che be called in case of querybypolygon
+     *
+     * @param active
+     */
+    const runDownload = async (active=false) => {
+
+      if (features.length > 1) {
+        layer[DownloadFormats.name].active = active;
+        this.setLayerActionTool({ layer });
+      }
+
+      const download_caller_id = ApplicationService.setDownload(true);
 
+      GUI.setLoadingContent(true);
+
+      try {
+        await CatalogLayersStoresRegistry.getLayerById(layer.id).getDownloadFilefromDownloadDataType(type, { data }) || Promise.resolve();
+      } catch(err) {
+        GUI.notify.error(err || t("info.server_error"));
+      }
+
+      ApplicationService.setDownload(false, download_caller_id);
+
+      GUI.setLoadingContent(false);
```

Comment:
@volterra79 after, try to see if you are able to rewrite these kind of statements to make use of some sort of download wrapper (eg. using some other `utils` function I'm not aware of):

```suggestion
await this.downloadApplicationWrapper(
        CatalogLayersStoresRegistry
          .getLayerById(layer.id)
          .getDownloadFilefromDownloadDataType
          .bind(this, type, { data })
      );
```